llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
+    Register Reg = DefMI->getOperand(0).getReg();
+
+    // Rematerialize MI to its use block. Since we are only rematerializing
+    // instructions that do not have any virtual reg uses, we do not need to
+    // call LiveRangeEdit::allUsesAvailableAt() and
+    // LiveRangeEdit::canRematerializeAt().
+    TII->reMaterialize(*InsertPos->getParent(), InsertPos, Reg,
+                       DefMI->getOperand(0).getSubReg(), *DefMI, *DAG.TRI);
+    MachineInstr *NewMI = &*std::prev(InsertPos);
+    NewMI->getOperand(0).setSubReg(DefMI->getOperand(0).getSubReg());
+    DAG.LIS->InsertMachineInstrInMaps(*NewMI);
+    NewMIs.push_back(NewMI);
+
+    // Update region boundaries in scheduling region we sinked from since we
+    // may sink an instruction that was at the beginning or end of its region.
+    DAG.updateRegionBoundaries(DAG.Regions, DefMI, /*NewMI =*/nullptr,
+                               /*Removing =*/true);
+
+    // Update region boundaries in region we sinked to.
+    DAG.updateRegionBoundaries(DAG.Regions, InsertPos, NewMI);
+
+    // Collect all regions impacted by the rematerialization and update their
+    // live-in/RP information.
+    for (unsigned I : Remat.LiveInRegions) {
+      ImpactedRegions.insert({I, false});
+      GCNRPTracker::LiveRegSet &RegionLiveIns = DAG.LiveIns[I];
+
+      // The register is no longer a live-in in all regions but the one that
+      // contains the single use. In live-through regions, maximum register
+      // pressure decreases predictably so we can directly update it. In the
+      // using region, maximum register pressure may or may not decrease, so we
+      // will mark it for re-computation after all materializations have taken
+      // place.
+      RegionLiveIns.erase(Reg);
+      if (Remat.UseMI->getParent() != DAG.Regions[I].first->getParent()) {
+        // Register is live-through and not used in this block.
+        LaneBitmask PrevMask = RegionLiveIns[Reg];
+        DAG.Pressure[I].inc(Reg, PrevMask, LaneBitmask::getNone(), DAG.MRI);
+      } else {
+        // Register is used in this block.
+        ImpactedRegions[I] = true;
+      }
     }
-    return false;
-  }
 
-  // Occupancy was improved for all regions.
-  for (auto &Entry : InsertedMIToOldDef) {
-    MachineInstr *MI = Entry.first;
-    MachineInstr *OldMI = Entry.second;
+    // RP in the region from which the instruction was rematerialized may or may
+    // not change, recompute-it fully later.
+    ImpactedRegions[Remat.DefRegion] = true;
+  }
 
-    // Remove OldMI from BBLiveInMap since we are sinking it from its MBB.
+  // Clean up the IR; remove rematerialized instructions from state.
+  for (auto [Remat, NewMI] : llvm::zip_equal(RematInstructions, NewMIs)) {
+    // Remove rematerialized instruction from BBLiveInMap since we are sinking
+    // it from its MBB. Also remove it from live intervals and from its MBB.
+    MachineInstr *OldMI = Remat.RematMI;
     DAG.BBLiveInMap.erase(OldMI);
```

Comment:
I don't think this is maintaining the map the way we want.

Reply:
After taking a closer look, it's probably not indeed (and potentially the same applies to `MBBLiveIns`). Working to fix this.

Reply:
Spent some more time looking at this, and it seems both `BBLiveInMap` and `MBBLiveIns` are not accessed after sinking instructions. Their only uses are in `computeBlockPressure` which is only called for other scheduling stages.

However, as I was trying to understand the exact purpose of `BBLiveInMap` I noticed that while `GCNScheduleDAGMILive::getRegionLiveInMap` seems to want to compute the live reg set at the beginning of the initial region of each block, the `BB` pointer is never updated. Seems like it should be inside the outer do/while loop.

https://github.com/llvm/llvm-project/blob/6780ab371ff8ce5e2b08b5fddd75a8296d3debdd/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp#L886-L894

What happens right now is that non-initial regions (if any) in the first block are skipped but they are not in other blocks. I don't see why the first block should be treated differently, is that a bug?